rust/agents/relayer/src/msg/metadata/multisig/weighted.rs

```diff
@@ -0,0 +1,88 @@
+use super::base::MultisigIsmMetadataBuilder;
+
+use crate::msg::metadata::{
+    multisig::{
+        MerkleRootMultisigMetadataBuilder, MessageIdMultisigMetadataBuilder, MetadataToken,
+        MultisigMetadata,
+    },
+    MessageMetadataBuilder,
+};
+use async_trait::async_trait;
+use derive_more::{AsRef, Deref};
+use derive_new::new;
+use eyre::{Context, Result};
+use hyperlane_base::MultisigCheckpointSyncer;
+use hyperlane_core::{HyperlaneMessage, H256};
+
+#[async_trait]
+pub trait WeightedMultisigIsmMetadataBuilder: MultisigIsmMetadataBuilder {
+    async fn ism_validator_requirements(
+        &self,
+        ism_address: H256,
+        message: &HyperlaneMessage,
+    ) -> Result<(Vec<(H256, u64)>, u64)> {
+        const CTX: &str = "When fetching WeightedMultisigIsm metadata";
+        let weighted_multisig_ism = self
+            .as_ref()
+            .build_weighted_multisig_ism(ism_address)
+            .await
+            .context(CTX)?;
+
+        let (validators, threshold_weight) = weighted_multisig_ism
+            .validators_and_threshold_weight(message)
+            .await
+            .context(CTX)?;
+
+        Ok((validators, threshold_weight))
+    }
+}
+
+#[derive(Debug, Clone, Deref, new, AsRef)]
+pub struct WeightedMerkleRootMultisigMetadataBuilder(MessageMetadataBuilder);
+
+#[async_trait]
+impl WeightedMultisigIsmMetadataBuilder for WeightedMerkleRootMultisigMetadataBuilder {}
+
+#[async_trait]
+impl MultisigIsmMetadataBuilder for WeightedMerkleRootMultisigMetadataBuilder {
+    fn token_layout(&self) -> Vec<MetadataToken> {
+        MerkleRootMultisigMetadataBuilder::new(self.0.clone()).token_layout()
+    }
+
+    async fn fetch_metadata(
+        &self,
+        validators: &[(H256, u64)],
+        threshold_weight: u64,
+        message: &HyperlaneMessage,
+        checkpoint_syncer: &MultisigCheckpointSyncer,
+    ) -> Result<Option<MultisigMetadata>> {
+        MerkleRootMultisigMetadataBuilder::new(self.0.clone())
+            .fetch_metadata(validators, threshold_weight, message, checkpoint_syncer)
+            .await
+    }
+}
```

Comment:
is there no way to just inherit the implementation?

Reply:
I've tried it but the compiler was complaining, don't remember the exact issue